pyop2/caching.py

```diff
@@ -513,10 +513,36 @@ def wrapper(*args, **kwargs):
                     debug(debug_string + "hit")
                     cache_hit = True
 
-                if configuration["spmd_strict"] and not pytools.is_single_valued(comm.allgather(cache_hit)):
-                    raise ValueError("Cache hit on some ranks but missed on others")
+                if isinstance(local_cache, DictLikeDiskAccess):
+                    if bcast:
+                        # Since disk caches share state between ranks there are extra
+                        # opportunities for mismatching hit/miss results and hence
+                        # deadlocks. These include:
+                        #
+                        # 1. Race conditions
+                        #
+                        # On CI or with ensemble parallelism other processes not in this
+                        # comm may write to disk, so load imbalances on the current comm
+                        # may result in a hit on some ranks but not others.
+                        #
+                        # 2. Eager writing to disk on rank 0
+                        #
+                        # Since broadcasting is non-blocking for the sending rank (rank 0)
+                        # it is possible for it to have written to disk before other ranks
+                        # begin the cache lookup. These ranks register a cache hit.
+                        #
+                        # If ranks disagree on whether it was a hit or miss then some ranks
+                        # will do a broadcast and others will not, ruining MPI synchronisation.
+                        # To fix this we make sure that in 'bcast' mode the hit/miss is
+                        # decided by rank 0.
+                        cache_hit = comm.bcast(cache_hit, root=0)
```

Comment:
I think this should be `comm.allreduce(cache_hit, op=MPI.LAND)`; if `cache_hit = True` on rank 0 and `cache_hit = False` on rank 1, `value` on rank 1 would never be updated (and causes some race condition at l.552).

Reply:
or maybe we could do:
```
root = rank if cache_hit else -1
root = comm.allreduce(root, op=MPI.MAX)
if root >=0:
    # bcast from root
```

Reply:
Ah yes of course. Excellent spot.

Reply:
I've done pretty much exactly what you've written here. Thanks so much!!

Reply:
Would parameterising the test on which rank is forced to wait in the line below make sure that we cover both cases - (rank 0 hits, rank 1 misses) or (rank 0 misses, rank 1 hits). https://github.com/firedrakeproject/firedrake/blob/7f59707d05909a33fa938914c7c5d787ac9129cf/tests/pyop2/test_caching.py#L907

Reply:
Sure. No harm in doing that.

Reply:
Done.